### PR TITLE
[IMP] chart: aggregate data for labels

### PR DESCRIPTION
**Pull request by yunchen4**
## Description:
This commit makes aggregating data for labels in the bar, line and pie chart possible, which makes more sense in practice. 

It's done by pre-processing the data before passing them into ChartJS to draw the plot. 

When a chart aggregates the data for labels, it will not be exported into .xlsx file, which is the same in GSheets. Excel currently doesn't support aggregating data for labels directly. 

Odoo task ID : [2832459](https://www.odoo.com/web#id=2832459&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_lt("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: src/components/side_panel/chart/bar_chart/bar_chart_config_panel.xml [hunk 1]
```diff
@@ -14,6 +14,18 @@
...
+        <div class="o-checkbox">
+          <label>
+            <input
+              type="checkbox"
+              name="aggregated"
+              t-att-checked="props.definition.aggregated"
+              t-on-change="onUpdateAggregated"
+              class="align-middle"
+            />
+            Aggregate labels
```
Comment (LucasLefevre) [inline src/components/side_panel/chart/bar_chart/bar_chart_config_panel.xml:26]: I'm not sure about the words "Aggregate labels"
Because you are not aggregating labels. You are *aggregating values based on labels**
What about just "Aggregate" ?

We could also add something like `title="Aggregate values with the same label"` on the label element

File: src/helpers/figures/charts/chart_ui_common.ts [hunk 2]
```diff
@@ -50,6 +50,33 @@ export function filterEmptyDataPoints(
...
 }
 
+export function aggregateDataForLabels(
+  labels: string[],
+  datasets: DatasetValues[]
+): { labels: string[]; dataSetsValues: DatasetValues[] } {
+  const parseNumber = (value) => (typeof value === "number" ? value : 0);
+
+  const labelMap = {};
+  for (let indexOfLabel = 0; indexOfLabel < labels.length; ++indexOfLabel) {
```
Comment (LucasLefevre) [inline src/helpers/figures/charts/chart_ui_common.ts:60]:
```suggestion
  for (const indexOfLabel of range(0, labels.length)) {
```
Slightly easier to read, more declarative, less moving parts.
`range` is inspired by the python `range` function :wink:

File: src/helpers/figures/charts/chart_ui_common.ts [hunk 3]
```diff
@@ -50,6 +50,33 @@ export function filterEmptyDataPoints(
...
+    const label = labels[indexOfLabel];
+    if (!labelMap[label]) {
+      labelMap[label] = datasets.map((dataset) => parseNumber(dataset.data[indexOfLabel]));
+      continue;
+    }
+    for (let indexOfDataset = 0; indexOfDataset < datasets.length; ++indexOfDataset) {
```
Comment (LucasLefevre) [inline src/helpers/figures/charts/chart_ui_common.ts:66]:
```suggestion
    for (const indexOfDataset of range(0, datasets.length)) {
```

File: src/helpers/figures/charts/chart_ui_common.ts [hunk 4]
```diff
@@ -50,6 +50,33 @@ export function filterEmptyDataPoints(
...
+  datasets: DatasetValues[]
+): { labels: string[]; dataSetsValues: DatasetValues[] } {
+  const parseNumber = (value) => (typeof value === "number" ? value : 0);
+
+  const labelMap = {};
+  for (let indexOfLabel = 0; indexOfLabel < labels.length; ++indexOfLabel) {
+    const label = labels[indexOfLabel];
+    if (!labelMap[label]) {
+      labelMap[label] = datasets.map((dataset) => parseNumber(dataset.data[indexOfLabel]));
+      continue;
```
Comment (LucasLefevre) [inline src/helpers/figures/charts/chart_ui_common.ts:64]: I would initialize `labelMap` for all labels first, with zeros.
And only then fill it with the values.
That would avoid this nested `if` and the `continue`

File: src/helpers/figures/charts/chart_ui_common.ts [hunk 5]
```diff
@@ -50,6 +50,33 @@ export function filterEmptyDataPoints(
   };
 }
 
+export function aggregateDataForLabels(
+  labels: string[],
+  datasets: DatasetValues[]
+): { labels: string[]; dataSetsValues: DatasetValues[] } {
+  const parseNumber = (value) => (typeof value === "number" ? value : 0);
+
+  const labelMap = {};
```
Comment (LucasLefevre) [inline src/helpers/figures/charts/chart_ui_common.ts:59]: add type annotation

File: src/helpers/figures/charts/chart_ui_common.ts [hunk 6]
```diff
@@ -50,6 +50,33 @@ export function filterEmptyDataPoints(
   };
 }
 
+export function aggregateDataForLabels(
```
Comment (LucasLefevre) [inline src/helpers/figures/charts/chart_ui_common.ts:53]: add a small docstring

File: src/xlsx/functions/charts.ts [hunk 7]
```diff
@@ -204,6 +204,13 @@ function addBarChart(chart: ExcelChartDefinition): XMLString {
   //
   // overlap and gapWitdh seems to be by default at -20 and 20 in chart.js.
   // See https://www.chartjs.org/docs/latest/charts/bar.html and https://www.chartjs.org/docs/latest/charts/bar.html#barpercentage-vs-categorypercentage
+
+  // Excel does not support aggregating labels
+  // GSheets does not export the chart if it aggregates the labels
+  if (chart.aggregated) {
```
Comment (LucasLefevre) [inline src/xlsx/functions/charts.ts:210]: how would you end up with a chart where aggregated is `true` ?
Aren't they filtered in `exportForExcel`?

File: src/types/chart/chart.ts [hunk 8]
```diff
@@ -57,6 +57,7 @@ export interface ExcelChartDefinition {
   readonly verticalAxisPosition: VerticalAxisPosition;
   readonly legendPosition: LegendPosition;
   readonly stacked?: boolean;
+  readonly aggregated?: boolean;
```
Comment (LucasLefevre) [inline src/types/chart/chart.ts:60]: if we can't export aggregated charts, `ExcelChartDefinition` should not have the `aggregated` key, right?

File: tests/plugins/chart/basic_chart.test.ts [hunk 9]
```diff
@@ -1446,6 +1448,216 @@ describe("Chart design configuration", () => {
...
+    let chart = (aggregatedModel.getters.getChartRuntime("42") as BarChartRuntime).chartJsConfig;
+    expect(chart.data!.datasets![0].data).toEqual([10, 11, 12, 13, 14, 15, 16, 17]);
+    expect(chart.data!.labels).toEqual(["P1", "P2", "P3", "P4", "P1", "P2", "P3", "P4"]);
+
+    updateChart(aggregatedModel, "42", { aggregated: true });
+    chart = (aggregatedModel.getters.getChartRuntime("42") as BarChartRuntime).chartJsConfig;
+    expect(chart.data!.datasets![0].data).toEqual([24, 26, 28, 30]);
+    expect(chart.data!.labels).toEqual(["P1", "P2", "P3", "P4"]);
+
+    updateChart(aggregatedModel, "42", { type: "line" });
```
Comment (LucasLefevre) [inline tests/plugins/chart/basic_chart.test.ts:1519]: you can use `jest.each`
```ts
  test.each(["bar", "line", "pie"] as const)("One dataset: all data complete", (type) => {
    createChart(aggregatedModel, aggregatedChart, "42");
    updateChart(aggregatedModel, "42", { type });
    let chart = (aggregatedModel.getters.getChartRuntime("42") as BarChartRuntime).chartJsConfig;
    expect(chart.data!.datasets![0].data).toEqual([10, 11, 12, 13, 14, 15, 16, 17]);
    expect(chart.data!.labels).toEqual(["P1", "P2", "P3", "P4", "P1", "P2", "P3", "P4"]);

    updateChart(aggregatedModel, "42", { aggregated: true });
    chart = (aggregatedModel.getters.getChartRuntime("42") as BarChartRuntime).chartJsConfig;
    expect(chart.data!.datasets![0].data).toEqual([24, 26, 28, 30]);
    expect(chart.data!.labels).toEqual(["P1", "P2", "P3", "P4"]);
  });
```

File: tests/plugins/chart/basic_chart.test.ts [hunk 10]
```diff
@@ -1446,6 +1448,216 @@ describe("Chart design configuration", () => {
...
+    expect(chart.data!.datasets![0].data).toEqual([10, 11, 12, 13, 14, 15, 16, 17]);
+    expect(chart.data!.labels).toEqual(["P1", "P2", "P3", "P4", "P1", "P2", "P3", "P4"]);
+
+    updateChart(aggregatedModel, "42", { aggregated: true });
+    chart = (aggregatedModel.getters.getChartRuntime("42") as BarChartRuntime).chartJsConfig;
+    expect(chart.data!.datasets![0].data).toEqual([24, 26, 28, 30]);
+    expect(chart.data!.labels).toEqual(["P1", "P2", "P3", "P4"]);
+  });
+
+  test("One dataset: some (not all) values are lost for a label", () => {
```
Comment (LucasLefevre) [inline tests/plugins/chart/basic_chart.test.ts:1540]: I don't understand this test title ?
What does "some but not all" mean ? Why are they lost ?
What does "lost" mean here ? Do you mean omitted ? ignored ?

What about something like `"empty cells are ignored when aggregating data"` ?

File: tests/plugins/chart/basic_chart.test.ts [hunk 11]
```diff
@@ -1446,6 +1448,216 @@ describe("Chart design configuration", () => {
...
+    setCellContent(aggregatedModel, "B3", "");
+    setCellContent(aggregatedModel, "B6", "");
+    chart = (aggregatedModel.getters.getChartRuntime("42") as BarChartRuntime).chartJsConfig;
+    expect(chart.data!.datasets![0].data).toEqual([10, 15, 28, 30]);
+    expect(chart.data!.labels).toEqual(["P1", "P2", "P3", "P4"]);
+
+    updateChart(aggregatedModel, "42", { aggregated: false });
+    chart = (aggregatedModel.getters.getChartRuntime("42") as BarChartRuntime).chartJsConfig;
+    expect(chart.data!.datasets![0].data).toEqual([10, undefined, 12, 13, undefined, 15, 16, 17]);
+    expect(chart.data!.labels).toEqual(["P1", "P2", "P3", "P4", "P1", "P2", "P3", "P4"]);
```
Comment (LucasLefevre) [inline tests/plugins/chart/basic_chart.test.ts:1560]: this is not related to aggregate and should be tested else where. (It is probably already tested)

File: tests/plugins/chart/basic_chart.test.ts [hunk 12]
```diff
@@ -1446,6 +1448,216 @@ describe("Chart design configuration", () => {
...
+    chart = (aggregatedModel.getters.getChartRuntime("42") as BarChartRuntime).chartJsConfig;
+    expect(chart.data!.datasets![0].data).toEqual([24, 26, 28, 30]);
+    expect(chart.data!.labels).toEqual(["P1", "P2", "P3", "P4"]);
+  });
+
+  test("One dataset: some (not all) values are lost for a label", () => {
+    createChart(aggregatedModel, aggregatedChart, "42");
+    let chart = (aggregatedModel.getters.getChartRuntime("42") as BarChartRuntime).chartJsConfig;
+    expect(chart.data!.datasets![0].data).toEqual([10, 11, 12, 13, 14, 15, 16, 17]);
+    expect(chart.data!.labels).toEqual(["P1", "P2", "P3", "P4", "P1", "P2", "P3", "P4"]);
```
Comment (LucasLefevre) [inline tests/plugins/chart/basic_chart.test.ts:1544]: don't repeat this everywhere.
Keep tests small and focused on the behavior under test.
The normal aggregate is already tested above

File: tests/plugins/chart/basic_chart.test.ts (hunk 10 above)
Comment (LucasLefevre) [inline tests/plugins/chart/basic_chart.test.ts:1540]: `"One dataset"` doesn't matter much for the test. I think it could be omitted.
Where it matters for the test is when testing with multiple datasets and that's where it's worth mentioning

File: tests/plugins/chart/basic_chart.test.ts [hunk 13]
```diff
@@ -1446,6 +1448,216 @@ describe("Chart design configuration", () => {
...
+    expect(chart.data!.datasets![0].data).toEqual([10, 15, 28, 30]);
+    expect(chart.data!.labels).toEqual(["P1", "P2", "P3", "P4"]);
+
+    updateChart(aggregatedModel, "42", { aggregated: false });
+    chart = (aggregatedModel.getters.getChartRuntime("42") as BarChartRuntime).chartJsConfig;
+    expect(chart.data!.datasets![0].data).toEqual([10, undefined, 12, 13, undefined, 15, 16, 17]);
+    expect(chart.data!.labels).toEqual(["P1", "P2", "P3", "P4", "P1", "P2", "P3", "P4"]);
+  });
+
+  test("One dataset: all values for a label are lost", () => {
```
Comment (LucasLefevre) [inline tests/plugins/chart/basic_chart.test.ts:1563]:
```suggestion
  test("label with no value aggregates to zero", () => {
```

File: tests/plugins/chart/basic_chart.test.ts [hunk 14]
```diff
@@ -1446,6 +1448,216 @@ describe("Chart design configuration", () => {
...
+    expect(chart.data!.datasets![0].data).toEqual([0, 26, 28, 30]);
+    expect(chart.data!.labels).toEqual(["P1", "P2", "P3", "P4"]);
+
+    updateChart(aggregatedModel, "42", { aggregated: false });
+    chart = (aggregatedModel.getters.getChartRuntime("42") as BarChartRuntime).chartJsConfig;
+    expect(chart.data!.datasets![0].data).toEqual([undefined, 11, 12, 13, undefined, 15, 16, 17]);
+    expect(chart.data!.labels).toEqual(["P1", "P2", "P3", "P4", "P1", "P2", "P3", "P4"]);
+  });
+
+  test("One dataset: some values are not numbers for a label", () => {
```
Comment (LucasLefevre) [inline tests/plugins/chart/basic_chart.test.ts:1586]:
```suggestion
  test("non number cells are ignored", () => {
```

File: tests/xlsx_export.test.ts [hunk 15]
```diff
@@ -759,6 +759,62 @@ describe("Test XLSX export", () => {
       expect(await exportPrettifiedXlsx(model)).toMatchSnapshot();
     });
 
+    test("charts that aggregate labels", async () => {
```
Comment (LucasLefevre) [inline tests/xlsx_export.test.ts:762]: I think you can just test `exportForExcel` and assert the charts are not exported.
No need to create (and commit) excel file which contains... nothing [interesting for that test]

**Review by LucasLefevre — COMMENTED** [review]
Nice job :)
You can use `jest.each` for all the tests instead of repeating the same lines again and again with only the type changed :wink:

File: src/components/side_panel/chart/bar_chart/bar_chart_config_panel.xml (hunk 1 above)
Comment (Francois-frgi) [reply src/components/side_panel/chart/bar_chart/bar_chart_config_panel.xml:26]: Yep agreed. It makes sense.

Comment (yunchen4) [reply src/components/side_panel/chart/bar_chart/bar_chart_config_panel.xml:26]: @Francois-frgi @LucasLefevre I'll add "Aggregate" checkbox just under categories/labels input (I saw in GSheet they place it in a similar way).

File: src/helpers/figures/charts/bar_chart.ts [hunk 16]
```diff
@@ -163,10 +167,12 @@ export class BarChart extends AbstractChart {
         : undefined,
       title: this.title,
       stacked: this.stacked,
+      aggregated: this.aggregated,
     };
   }
 
-  getDefinitionForExcel(): ExcelChartDefinition {
+  getDefinitionForExcel(): ExcelChartDefinition | undefined {
+    if (this.aggregated) return undefined;
```
Comment (LucasLefevre) [inline src/helpers/figures/charts/bar_chart.ts:176]: add a comment to explain why

File: src/helpers/figures/charts/line_chart.ts [hunk 17]
```diff
@@ -176,7 +181,8 @@ export class LineChart extends AbstractChart {
     return new LineChart(definition, this.sheetId, this.getters);
   }
 
-  getDefinitionForExcel(): ExcelChartDefinition {
+  getDefinitionForExcel(): ExcelChartDefinition | undefined {
+    if (this.aggregated) return undefined;
```
Comment (LucasLefevre) [inline src/helpers/figures/charts/line_chart.ts:186]: same

File: src/helpers/figures/charts/pie_chart.ts [hunk 18]
```diff
@@ -164,7 +169,8 @@ export class PieChart extends AbstractChart {
     return new PieChart(definition, sheetId, this.getters);
   }
 
-  getDefinitionForExcel(): ExcelChartDefinition {
+  getDefinitionForExcel(): ExcelChartDefinition | undefined {
+    if (this.aggregated) return undefined;
```
Comment (LucasLefevre) [inline src/helpers/figures/charts/pie_chart.ts:174]: same

**Review by LucasLefevre — APPROVED** [review]
Make sure the lines in the commit message are under 80 characters
I already applied the comments :wink: 

robodoo r+
